tree/dataframe/inc/ROOT/RDF/RInterface.hxx

```diff
@@ -2232,31 +2233,36 @@ public:
    ///
    ColumnNames_t GetColumnNames()
    {
-      ColumnNames_t allColumns;
+      // there could be duplicates between Redefined columns and columns in the data source
+      std::unordered_set<std::string> allColumns;
 
       auto addIfNotInternal = [&allColumns](std::string_view colName) {
          if (!RDFInternal::IsInternalColumn(colName))
-            allColumns.emplace_back(colName);
+            allColumns.emplace(colName);
       };
 
-      auto columnNames = fColRegister.GetNames();
+      auto definedColumns = fColRegister.GetNames();
 
-      std::for_each(columnNames.begin(), columnNames.end(), addIfNotInternal);
+      std::for_each(definedColumns.begin(), definedColumns.end(), addIfNotInternal);
 
       auto tree = fLoopManager->GetTree();
       if (tree) {
          auto branchNames = RDFInternal::GetBranchNames(*tree, /*allowDuplicates=*/false);
-         allColumns.insert(allColumns.end(), branchNames.begin(), branchNames.end());
+         for (const auto &bName : branchNames)
+            allColumns.emplace(bName);
       }
 
       if (fDataSource) {
          const auto &dsColNames = fDataSource->GetColumnNames();
-         // ignore columns starting with R_rdf_sizeof_
-         std::copy_if(dsColNames.begin(), dsColNames.end(), std::back_inserter(allColumns),
-                      [](const std::string &s) { return s.rfind("R_rdf_sizeof", 0) != 0; });
+         for (const auto &s : dsColNames) {
+            if (s.rfind("R_rdf_sizeof", 0) != 0)
+               allColumns.emplace(s);
+         }
       }
 
-      return allColumns;
+      ColumnNames_t ret(allColumns.begin(), allColumns.end());
+      std::sort(ret.begin(), ret.end());
```

Comment:
The fact that we are sorting the returned column names is also new, right?

Reply:
yes, the order was and is unspecified anyway, but alphabetical ordering reads better

Reply:
Ok, since it's a change in the behaviour of the method and it's not related to avoiding duplicates you might want to have it in a separate commit then.

Reply:
using an unordered set already changes the order, i'm just picking a new order that's better than what unordered_set provides -- but ok

Reply:
done